File: Python/insercion_centinela.py
# Programa para buscar un elemento con centinela y ordenar la lista con insertion sort

# Funcion para llenar una lista

def fill_list(lista):
    length = int(input("\nIngrese la cantidad de elementos de la lista: "))

    for i in range(0, length):
        value = int(input(f"Ingrese el elemento numero {i}: "))
        lista.append(value)

# Funcion de busqueda con centinela

def sentinel_search(lista, size, v):
    lista.append(v)
    i = 0

    while (lista[i] != v):
        i += 1
    
    if i < size:
        return i
    else:
        return -1

# Funcion para el insertion sort

def insertion_sort(lista, size):

    for i in range(1, size):
        t = lista[i]
        j = i
        while (j > 0) and (t < lista[j - 1]):
            lista[j] = lista[j - 1]
            j -= 1
        
        lista[j] = t



# Programa main()

int_list = []

# Llamada a funcion para llenar la lista

fill_list(int_list)
print(f"\nLa lista es: {int_list}")
list_size = len(int_list)

# Llamada a funcion para buscar un elemento con centinela

value = int(input("Ingrese el valor del elemento a buscar: "))
pos = sentinel_search(int_list, list_size, value)
print(f"\nLa posicion del elemento {value} es: {pos}")
int_list.pop()

# Llamada a funcion para ordenar la lista mediante insertion sort

print("\nOrdenando por insertion sort...")
insertion_sort(int_list, list_size)
print(f"\nLa lista es: {int_list}")

File: C/insercion_centinela.c
// Programa para implementar la busqueda con centinela y el ordenamiento por insertion sort

#include <stdio.h>

// Funcion para llenar un arreglo

void fill_array(int arr[], int size)
{
    register int i;
    int element;

    for (i = 0; i < size; i++)
    {
        printf("\nIngrese el valor del elemento %d ", i);
        scanf("%d", &element);
        arr[i] = element;
    }
    printf("\n");
}

// Funcion para imprimir un arreglo

void print_array(int arr[], int size)
{
    register int i;

    printf("\nLos elementos del arreglo son: [ ");
    
    for (i = 0; i < size; i++)
    {
        printf("%d ",arr[i]);
    }
    
    printf("]\n");
}

// Funcion para buscar un elemento por centinela

int sentinel_search(int arr[], int size, int v)
{
    register int i;

    for (i = 0, arr[size] = v; arr[i] != v; i++);

    if (i < size)
        return i;
    else
        return -1;
}

// Funcion para el insertion sort

void insertion_sort(int arr[], int size)
{
    register int i, j;
    int t;

    for (i = 1; i < size; i++)
    {
        t = arr[i];
        for (j = i; j > 0 && t < arr[j - 1]; j--)
            arr[j] = arr[j - 1];

        arr[j] = t;
    }
}

void main()
{
    int n, value, pos;

    printf("\nIngrese la cantidad de elementos del arreglo: ");
    scanf("%d", &n);

    int array[n];

    // Llamada a funcion para llenar el arreglo

    fill_array(array, n);
    print_array(array, n);

    // Llamada a funcion para buscar un elemento por centinela

    printf("\nIngrese el valor del elemento a buscar: ");
    scanf("%d", &value);

    pos = sentinel_search(array, n, value);
    printf("\nLa posicion del elemento %d es: %d", value, pos);

    // Llamada a funcion para el ordenamiento por insertion sort

    printf("\nOrdenando por insertion sort!");
    insertion_sort(array, n);
    print_array(array, n);
}